io/private/copcwriter/Output.cpp

```diff
@@ -0,0 +1,352 @@
+            else
+                m_header.points_by_return[i - 1] = 0;
+        }
+    }
+
+    if (m_header.point_count_14 > (std::numeric_limits<uint32_t>::max)())
+        m_header.point_count = 0;
+
+    // Note that only these VLRs go in the standard VLR area. The rest are written as
+    // extended VLRs.
+    setupVlrs();
+    m_header.point_offset = lazperf::header14::Size + m_vlrBuf.size();
+
+    // The chunk table offset is written as the first 8 bytes of the point data in LAZ.
+    m_chunkOffsetPos = m_header.point_offset;
+
+    // The actual point data comes after the chunk table offset. m_pointPos is updated
+    // as points are written.
+    m_pointPos = m_chunkOffsetPos + sizeof(uint64_t);
+
+    m_f.open(b.opts.filename, std::ios::out | std::ios::binary);
+}
+
+// For this to work properly, the VLRs can't change size.
+void Output::setupVlrs()
+{
+    m_vlrBuf.clear();
+    m_header.vlr_count = 0;
+    addCopcVlr();
+    addLazVlr();
+    addWktVlr();
+    addEbVlr();
+}
+
+void Output::addCopcVlr()
+{
+    // The copc VLR isn't local because we need to fill in the hierarchy information
+    // when we know it.
+    m_copcVlr.center_x = (b.bounds.maxx / 2) + (b.bounds.minx / 2);
+    m_copcVlr.center_y = (b.bounds.maxy / 2) + (b.bounds.miny / 2);
+    m_copcVlr.center_z = (b.bounds.maxz / 2) + (b.bounds.minz / 2);
+    m_copcVlr.halfsize = (b.bounds.maxx - b.bounds.minx) / 2;
+    m_copcVlr.spacing = 2 * m_copcVlr.halfsize / CellCount;
+    m_copcVlr.gpstime_minimum = b.stats[(int)stats::Index::GpsTime].minimum();
+    m_copcVlr.gpstime_maximum = b.stats[(int)stats::Index::GpsTime].maximum();
+
+    std::vector<char> buf = m_copcVlr.header().data();
+    m_vlrBuf.insert(m_vlrBuf.end(), buf.begin(), buf.end());
+
+    buf = m_copcVlr.data();
+    m_vlrBuf.insert(m_vlrBuf.end(), buf.begin(), buf.end());
+    m_header.vlr_count++;
+}
+
+void Output::addLazVlr()
+{
+    lazperf::laz_vlr laz(b.pointFormatId, b.numExtraBytes, lazperf::VariableChunkSize);
+
+    std::vector<char> buf = laz.header().data();
+    m_vlrBuf.insert(m_vlrBuf.end(), buf.begin(), buf.end());
+
+    buf = laz.data();
+    m_vlrBuf.insert(m_vlrBuf.end(), buf.begin(), buf.end());
+    m_header.vlr_count++;
+}
+
+void Output::addWktVlr()
+{
+    if (!b.srs.valid())
+        return;
+
+    lazperf::wkt_vlr wkt(b.srs.getWKT1());
+
+    std::vector<char> buf = wkt.header().data();
+    m_vlrBuf.insert(m_vlrBuf.end(), buf.begin(), buf.end());
+
+    buf = wkt.data();
+    m_vlrBuf.insert(m_vlrBuf.end(), buf.begin(), buf.end());
+    m_header.vlr_count++;
+}
+
+void Output::addEbVlr()
+{
+    if (!b.numExtraBytes)
+        return;
+
+    lazperf::eb_vlr eb(b.numExtraBytes);
+
+    std::vector<char> buf = eb.header().data();
+    m_vlrBuf.insert(m_vlrBuf.end(), buf.begin(), buf.end());
+
+    buf = eb.data();
+    m_vlrBuf.insert(m_vlrBuf.end(), buf.begin(), buf.end());
+    m_header.vlr_count++;
+}
```

Comment:
We also need the ability to add user-defined VLRs and #3683

Reply:
"Need" is a strong word here.